src/ddscxx/include/dds/sub/detail/SamplesHolder.hpp

```diff
@@ -139,8 +139,39 @@ class CDRSamplesHolder : public SamplesHolder
       uint32_t cpp_sample_size = this->samples_.delegate()->length();
       for (uint32_t i = 0; i < cpp_sample_size; ++i)
       {
+        struct ddsi_serdata * current_blob = cdr_blobs[i];
+#ifdef DDS_HAS_SHM
+        // if the data is over iox_chunk
+        if (current_blob->iox_chunk && current_blob->iox_subscriber) {
+          // get the data
+          org::eclipse::cyclonedds::topic::CDRBlob & sample_data =
+            (*this->samples_.delegate())[i].delegate().data();
+          // update the header
+          memcpy(sample_data.encoding().data(), current_blob->iox_chunk, 4);
+          // update the data kind
+          sample_data.kind(
+            static_cast<org::eclipse::cyclonedds::topic::BlobKind>(current_blob->kind));
+          if (sample_data.kind() != org::eclipse::cyclonedds::topic::BlobKind::Empty) {
+            // get the size
+            auto iox_header = iceoryx_header_from_chunk(current_blob->iox_chunk);
+            // if the iox chunk has the data in serialized form
+            if (iox_header->shm_data_state == IOX_CHUNK_CONTAINS_SERIALIZED_DATA) {
```

Comment:
The state can also now be obtained with `iox_shm_data_state_t shm_get_data_state(void *iox_chunk);` Recommended if the header is not required for anything else.

Reply:
I need the header for the size of the chunk, so I was not using `shm_get_data_state()`


Ideally, the interface might need to have proper setters and getters for the fields in the user header, we can probably do as a followup